daft/runners/distributed_swordfish.py

```diff
@@ -0,0 +1,194 @@
+        concated = MicroPartition.concat(list(partitions))
+        return PartitionMetadata.from_table(concated), concated
+
+
+class RaySwordfishTaskHandle:
+    """RaySwordfishTaskHandle is a handle to a task that is running on a swordfish actor.
+
+    It is used to asynchronously get the result of the task, cancel the task, and perform any post-task cleanup.
+    """
+
+    def __init__(
+        self,
+        result_handle: ray.ObjectRef,
+        actor_handle: ray.actor.ActorHandle,
+        add_memory_callback: Callable[[int], None],
+        task_memory_cost: int,
+    ):
+        self.result_handle = result_handle
+        self.actor_handle = actor_handle
+        self.add_memory_callback = add_memory_callback
+        self.task_memory_cost = task_memory_cost
+
+    async def _get_result(self) -> RayPartitionRef:
+        results = []
+        async for result in self.result_handle:
+            results.append(result)
+
+        metadata, result = self.actor_handle.concat_and_get_metadata.options(num_returns=2).remote(*results)
+        metadata = await metadata
+        self.add_memory_callback(self.task_memory_cost)
+        return RayPartitionRef(result, metadata.num_rows, metadata.size_bytes)
+
+    async def get_result(self) -> RayPartitionRef:
+        task = asyncio.create_task(self._get_result())
+        return await task
+
+    def cancel(self):
+        ray.cancel(self.result_handle)
+
+
+class RaySwordfishWorkerHandle:
+    """RaySwordfishWorkerHandle is a wrapper around a ray swordfish actor.
+
+    It is used to submit tasks to the worker and keep track of the worker's node id, handle, number of cpus, total and available memory.
+    """
+
+    def __init__(
+        self,
+        actor_node_id: str,
+        actor_handle: ray.actor.ActorHandle,
+        num_cpus: int,
+        total_memory_bytes: int,
+    ):
+        self.actor_node_id = actor_node_id
+        self.actor_handle = actor_handle
+        self.num_cpus = num_cpus
+        self.total_memory_bytes = total_memory_bytes
+        self.available_memory_bytes = total_memory_bytes
+
+    def get_actor_node_id(self) -> str:
+        return self.actor_node_id
+
+    def get_num_cpus(self) -> int:
+        return self.num_cpus
+
+    def get_total_memory_bytes(self) -> int:
+        return self.total_memory_bytes
+
+    def get_available_memory_bytes(self) -> int:
+        return self.available_memory_bytes
+
+    def add_back_memory(self, memory_bytes: int):
+        self.available_memory_bytes += memory_bytes
+
+    def submit_task(self, task: RaySwordfishTask) -> RaySwordfishTaskHandle:
+        self.available_memory_bytes -= task.estimated_memory_cost()
+        psets = {k: [v.object_ref for v in v] for k, v in task.psets().items()}
+        return RaySwordfishTaskHandle(
+            self.actor_handle.run_plan.remote(task.plan(), psets, task.execution_config()),
```

Comment:
we can leverage a done callback to add back the memory rather than doing it when `get_result`

https://docs.python.org/3/library/asyncio-future.html#asyncio.Future.add_done_callback